kernel/src/common/deferred_call_mux.rs

```diff
@@ -0,0 +1,207 @@
+//!
+//! These are scheduled "interrupts" in the chip scheduler.
+//! They are especially important if some hardware doesn't
+//! support real interrupts where they are needed, or for
+//! implementing software devices that are supposed to work
+//! like hardware devices.
+//!
+//! Usage
+//! -----
+//!
+//! The `deferred_call_mux_clients` array size determines how many
+//! [DeferredCallHandle](crate::common::deferred_call_mux::DeferredCallHandle)s
+//! may be registered with the Mux.
+//! When no more slots are available,
+//! `deferred_call_mux.register(some_client)` will return `None`.
+//!
+//! ```
+//! # use core::cell::Cell;
+//! # use kernel::common::cells::OptionalCell;
+//! # use kernel::static_init;
+//! use kernel::common::deferred_call_mux::{
+//!     DeferredCallMux,
+//!     DeferredCallMuxClient,
+//!     set_global_mux,
+//! };
+//!
+//! let deferred_call_mux_clients = unsafe { static_init!(
+//!     [(Cell<bool>, OptionalCell<&'static DeferredCallMuxClient>); 1],
+//!     [(Cell::new(false), OptionalCell::empty())]
+//! ) };
+//! let deferred_call_mux = unsafe { static_init!(
+//!     DeferredCallMux,
+//!     DeferredCallMux::new(deferred_call_mux_clients)
+//! ) };
+//! assert!(unsafe { set_global_mux(deferred_call_mux) }, true);
+//!
+//! # struct SomeCapsule;
+//! # impl SomeCapsule {
+//! #     pub fn new(_mux: &'static DeferredCallMux) -> Self { SomeCapsule }
+//! #     pub fn set_deferred_call_handle(
+//! #         &self,
+//! #         _handle: kernel::common::deferred_call_mux::DeferredCallHandle,
+//! #     ) { }
+//! # }
+//! # impl DeferredCallMuxClient for SomeCapsule {
+//! #     fn call(
+//! #         &self,
+//! #         _handle: kernel::common::deferred_call_mux::DeferredCallHandle,
+//! #     ) { }
+//! # }
+//! #
+//! // Here you can register custom capsules, etc.
+//! // This could look like:
+//! let some_capsule = unsafe { static_init!(
+//!     SomeCapsule,
+//!     SomeCapsule::new(deferred_call_mux)
+//! ) };
+//! some_capsule.set_deferred_call_handle(
+//!     deferred_call_mux.register(some_capsule).expect("no deferred call slot available")
+//! );
+//! ```
+
+use crate::common::cells::OptionalCell;
+use crate::common::deferred_call::DeferredCall;
+use core::cell::Cell;
+
+pub const DEFERRED_CALL_MUX_TASK: usize = 31; // highest bit in usize reserved for deferred call mux
+
+static mut DEFERRED_CALL_MUX: Option<&'static DeferredCallMux> = None;
+
+/// Sets a global `DeferredCallMux` instance
+///
+/// This is required before any deferred calls can be retrieved.
+/// It may be called only once. Returns `true` if the mux was successfully
+/// registered.
+pub unsafe fn set_global_mux(mux: &'static DeferredCallMux) -> bool {
+    // If the returned reference is identical to the mux argument,
+    // it is set in the option. Otherwise, a different mux is
+    // already registered and may not be replaced.
+    (*DEFERRED_CALL_MUX.get_or_insert(mux)) as *const _ == mux as *const _
```

Comment:
I'm not quite sure whether it's a good idea to prevent setting a global `Mux` a second time. While this could result in a `Mux` getting called even though it hasn't request a deferred call, nothing too bad should happen if the global Mux gets changed.

Reply:
I don't understand why you would replace it or need to configure it explicitly in separate method instead of doing this in the constructor. 

However, the docs are really good :+1:

Reply:
I don't know how it would be possible for me to configure the global `Mux` reference in the constructor, as I can't create a reference that persists while the struct is still being moved to it's final destination, right?

I assumed that it is best practice to `static_init!` only in the `boards`-crate, where all other capsules are initialized too. So the `set_global_mux` function serves a somewhat similar purpose as all the `set_client`s.

Reply:
Right, I think I got confused and mixed up `DeferredCallMux` and `DeferredCallMuxClient` sorry!

Reply:
No worries. I think my naming is a bit confusing and way to _enterprisified_. In general I'm having a hard time coming up with good names for all the structs, so suggestions are welcome. :laughing: